### PR TITLE
feat: add `is_parameter_timeseries` trait, improve parameter timeseries indexing

### PR DESCRIPTION
**Pull request by AayushSabharwal**
## Checklist

- [ ] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API

### CODE REVIEW

File: docs/src/api.md [hunk 1]
```diff
@@ -30,10 +30,39 @@ allvariables
...
 ```@docs
 observed
+parameter_observed
+ParameterObservedFunction
+```
+
+#### Parameter timeseries
+
+If the index provider contains parameters that change during the course of the simulation
+at discrete intervals, it must implement the following methods to ensure correct
```
Comment (baggepinnen) [inline docs/src/api.md:40]:
```suggestion
at discrete time points, it must implement the following methods to ensure correct
```

File: docs/src/api.md [hunk 2]
```diff
@@ -30,10 +30,39 @@ allvariables
...
+functioning of [`getu`](@ref) and [`getp`](@ref) for value providers that save the parameter
+timeseries. Note that there can be multiple parameter timeseries, in case different parameters
+change at different times.
```
Comment (baggepinnen) [inline docs/src/api.md:43]:
```suggestion
may change at different times.
```

File: docs/src/complete_sii.md [hunk 3]
```diff
@@ -305,6 +313,164 @@ integrator.ps[:b] = 3.0
 setp(integrator, :b)(integrator, 3.0) # functionally the same as above
 ```
 
+## Parameter Timeseries
+
+If a solution object saves modified parameter values (such as through callbacks) during the
```
Comment (baggepinnen) [inline docs/src/complete_sii.md:318]:
```suggestion
If a solution object includes modified parameter values (such as through callbacks) during the
```

Comment (AayushSabharwal) [reply docs/src/complete_sii.md]: Fixed in https://github.com/SciML/SymbolicIndexingInterface.jl/pull/75/commits/9046cd71a4147ac48322914436f422f847c4624f

File: docs/src/api.md (hunk 2 above)
Comment (AayushSabharwal) [reply docs/src/api.md]: Fixed in https://github.com/SciML/SymbolicIndexingInterface.jl/pull/75/commits/9046cd71a4147ac48322914436f422f847c4624f

File: docs/src/api.md (hunk 1 above)
Comment (AayushSabharwal) [reply docs/src/api.md]: Fixed in

**Comment by AayushSabharwal** [comment]
I'll reopen to run CI once MTK is tagged

**Comment by AayushSabharwal** [comment]
RAT downstream failure is an adjoint bug in SciMLBase

**Comment by AayushSabharwal** [comment]
``

**Comment by AayushSabharwal** [comment]
SciMLBase downstream failure is an MTK bug

**Pull request merged**